wireguard: Remove operator and disable KPR encryption

brb opened the pull request:
See commit msgs.

Fix https://github.com/cilium/cilium/issues/15547.

brb committed:
wireguard: Remove operator

This commit removes the Wireguard operator which previously was used to
allocate a Wireguard tunnel IP.

After taking a closer look into the Wireguard source code in the kernel
and realising that we don't see the tunnel IP on the wire, we came to
realisation that the IP is not needed at all.

For host2host encryption with Wireguard (TODO) we don't need to rely on
the tunnel IP (we will be able to use the nodeIP and Wireguard's fwmark
to avoid the loop).

Signed-off-by: Martynas Pumputis <m@lambda.lt>

brb committed:
wireguard: Completely disable rp_filter

Just having a default route is not enough to pass the rp_filter check in
the loose mode. So, disable it completely.

Signed-off-by: Martynas Pumputis <m@lambda.lt>

borkmann reviewed (approved):
Nice, less knobs the better!

gandro reviewed (changes requested):
:tada: Good riddance! 

I believe this can also be removed:
https://github.com/cilium/cilium/blob/81b2ac0e51a6a22a64388babeb6f09ea6b17d382/pkg/wireguard/types/types.go#L25-L31

twpayne reviewed (approved):
👍

tklauser commented on pkg/wireguard/agent/agent.go:
```diff
@@ -24,7 +24,7 @@ import (
 	"strconv"
 
-	"github.com/cilium/cilium/pkg/cidr"
+	//"github.com/cilium/cilium/pkg/cidr"
```
Remove this instead of commenting it out?

tklauser commented on pkg/wireguard/agent/agent.go:
```diff
@@ -48,26 +48,22 @@ const (
-// Once RestoreFinished() is called obsolete keys and peers are removed.
-// UpdatePeer() inserts or updates the public key of peer discovered via
-// the node manager.
+// Agent needs to be initialized with Init()In .Init(), the Wireguard tunnel
```
```suggestion
// Agent needs to be initialized with Init(). In .Init(), the Wireguard tunnel
```

tklauser reviewed (approved):
Nice :rocket:

gandro commented on pkg/wireguard/agent/agent.go:
```diff
@@ -242,10 +196,8 @@ func (a *Agent) UpdatePeer(nodeName, pubKeyHex string,
 		logfields.PubKey:    pubKeyHex,
 		logfields.NodeIPv4:  nodeIPv4,
 		logfields.PodCIDRv4: podCIDRv4,
-		logfields.WgIPv4:    wgIPv4,
```
You also want to remove the constant from the `logfields` package.

brb replied:
You are right again!

christarazi reviewed (approved):
:rocket:

pchaigno reviewed (approved):
Do we need a new codeowner for `/pkg/wireguard/`?

borkmann commented on Documentation/cmdref/cilium-agent.md:
```diff
@@ -42,6 +42,7 @@ cilium-agent [flags]
       --bpf-lb-mode string                                   BPF load balancing mode ("snat", "dsr", "hybrid") (default "snat")
       --bpf-lb-rss-ipv4-src-cidr string                      BPF load balancing RSS outer source IPv4 CIDR prefix for IPIP
       --bpf-lb-rss-ipv6-src-cidr string                      BPF load balancing RSS outer source IPv6 CIDR prefix for IPIP
+      --bpf-lb-wireguard                                     Enable Wireguard encryption for packets forwarded by Cilium LB (e.g. NodePort)
```
Just a thought: Couldn't we pick a reasonable default without delegating extra opt in/out for users? How is the encryption handled in case of IPsec for (kube-proxy) NodePort traffic where the backend is remote? Couldn't we default to the same behavior and document that? What if in the load-balancer there are some services where backends are outside of K8s scope, and thus have no context of wireguard? Probably it makes sense to stick to encryption by default for backends in the ipcache, and for everything external (aka world) to have no encryption. The east-west case with sock LB will presumably also stick to encryption, so north-south case could be consistent to that. Otoh, one could argue if we're just punting client packets from A to B they might have arrived at B in the first place, so what would be the point of encryption there, but it may be weird given these have the src IP of the cluster node which emit pkts in clear text.

brb replied:
> Couldn't we pick a reasonable default without delegating extra opt in/out for users? 

Yeah, I was thinking about it, but my motivation for introducing the flag is the following:
1. Some users might not want to have the asymmetry (pod2host => cilium_wg0, host2pod=> eth0) introduced by the KPR encryption.
2. Not all users might need the KPR encryption.
3. Users who are running older than 5.8 kernel and has KPR enabled won't be able to enable Wireguard due to the absent `bpf_skb_change_head()` helper.

> How is the encryption handled in case of IPsec for (kube-proxy) NodePort traffic where the backend is remote?

Haven't tested, but I expect it to encrypt the traffic if the host2host encryption is enabled.

> What if in the load-balancer there are some services where backends are outside of K8s scope, and thus have no context of wireguard?

Such users would use `--bpf-lb-wireguard=false`.

> Probably it makes sense to stick to encryption by default for backends in the ipcache, and for everything external (aka world) to have no encryption.

Yep, but unfortunately we currently don't encrypt host2pod traffic (non-NodePort).

> The east-west case with sock LB will presumably also stick to encryption, so north-south case could be consistent to that.

Yep, that's the case today (except when a client is running host netns).

brb replied:
Discussed offline. For now until we have host2pod encryption let's disable KPR encryption. /cc @gandro

brb committed:
wireguard: Disable KPR encryption

After removing the Wireguard tunnel IP, a traffic forwarded by NodePort
BPF cannot be encrypted as it used to be.

When running Wireguard with KPR, we want to encrypt the following
NodePort BPF traffic:

    1. A request from a client to a service handled by NodePort BPF when
       a remote service endpoint is selected (after SNAT and DNAT).
    2. A reply from the service endpoint to the request above from on
       the remote node.

The 1. is marked for encryption from the NodePort BPF program (once it
detects that the service endpoint is remote). The encryption of 2. is
more complicated.

Previously, when we had a Wireguard tunnel IP, the request was SNAT-ed
to the tunnel IP. This made the reply to automatically go over the
Wireguard tunnel.  Unfortunately, after removing the Wireguard tunnel
IP, this is no longer the case. The request is SNAT-ed to a node IP, and
therefore a reply (src=podIP|dst=nodeIP) is no longer encrypted, as it
does not go over the Wireguard tunnel.

To fix this, we could mark a packet sent from a pod for the encryption
if we detect that dst is REMOTE_NODE_ID. Then later on, the packet could
be routed by the stack to the Wireguard tunnel device (thanks to the
mark).  Unfortunately, this introduces an asymmetry for non-NodePort
host2pod traffic, as host2pod will go unencrypted over a native
device while pod2host will go over the tunnel and will be encrypted.

Because of the asymmetry, disable the KPR encryption until we can
encrypt host2pod with Wireguard. In the end, the NodePort BPF encryption
is the host2pod and pod2host encryption.

Signed-off-by: Martynas Pumputis <m@lambda.lt>

pchaigno commented on pkg/wireguard/agent/agent.go:
```diff
@@ -48,26 +47,22 @@ const (
-// Once RestoreFinished() is called obsolete keys and peers are removed.
-// UpdatePeer() inserts or updates the public key of peer discovered via
-// the node manager.
+// Agent needs to be initialized with Init()In. Init(), the Wireguard tunnel
```
Small typo here. Probably not worth pushing if it's the only thing left.
